### ISSUE TITLE
Use same sorting logic for sorting dependencies as npm v7

### ISSUE THREAD

**Issue by G-Rath** [keithamus/sort-package-json#234]
NPM v7 changed the way dependencies are sorted very slightly which means this:

```
{
  "name": "spj-te",
  "dependencies": {
    "sort-package-json": "^1.52.0"
  },
  "devDependencies": {
    "@types/koa-bodyparser": "^4.3.3",
    "@types/koa__router": "^8.0.8"
  }
}
```

becomes this:

```
{
  "name": "spj-te",
  "dependencies": {
    "sort-package-json": "^1.52.0"
  },
  "devDependencies": {
    "@types/koa__router": "^8.0.8",
    "@types/koa-bodyparser": "^4.3.3"
  }
}
```

The specific change they made was to use `localeCompare`, which technically the better way for sorting strings than just calling `sort` but means npm v7 & above sort differently compared to both `sort-package-json` and the other two main package managers.

I opened an issue for discussion about this but it was closed with a comment about the change being intentional: https://github.com/npm/cli/issues/3935

I was thinking that `sort-package-json` could attempt to figure out if it should use `localeCompare` by trying to read `package-lock.json` and checking if the `lockfileVersion` is greater than 1, otherwise falling back to its current behaviour if its 1 or the lock cannot be read.

I'm happy to implement this change.

**Comment by Nokel81** [comment]
Any update on this? Would be nice to have the dependencies sorted the same order as `npm` has them

**Comment by G-Rath** [comment]
@Nokel81 I've been really on the fence about this because [NPM v7+ actually sorts slightly different to all other package managers](https://github.com/npm/cli/issues/3935) and [they're not going to change that](https://github.com/npm/package-json/pull/2#issuecomment-1012267763) because apparently it'll create too much churn for NPM v7+ users...

In my experience I've only ever had this once on a project, with the dependencies mentioned above since it only impacts the sort order of `-` vs `_`, which is very rare to have packages matching except for those characters.

**Comment by Nokel81** [comment]
I noticed it specifically because of the dash and underscore. Due to `ansi_up` and `ansi-exp`.

I definitely think that fit projects using `npm` should be able to specify this sort order

**Comment by keithamus** [comment]
We could also detect what npm version is being used and do the right thing.

I'm happy to look over any PR that anyone wants to write on this :wink:

**Comment by G-Rath** [comment]
Actually with [`packageManager`](https://nodejs.org/api/packages.html#packagemanager) being in `package.json`, that should be very easy to do 🤔

**Comment by rgant** [comment]
Mark this to look into working on later as it annoys me so much that `eslint_d` gets moved around by `sort-package-json` and `npm install`.

**Comment by G-Rath** [comment]
Yeah I've actually had this tab open for a few months to get done when I have some bandwidth, as we've recently started switching everything over to using npm at work 😅

fwiw at this point I think it should be lockfile based i.e. applying the different sort only if `package-lock.json` is present.

**Comment by karlhorky** [comment]
I opened a preliminary PR here, prompted by my issue in #355 :

- #358
